### DISCUSSION TITLE
Issue with Missing Dimnames and Assay Class Change in Seurat v5.0.1

### DISCUSSION THREAD

**Discussion by qingmemng** [satijalab/seurat#8234] [Q&A]
![屏幕截图 2023-12-28 154429](https://github.com/satijalab/seurat/assets/143380773/d8de6b42-141a-4c4c-94c6-edb45a85ebc7)
Hi!

I recently started working with Seurat version 5.0.1 for single-cell RNA-seq analysis. While executing my R code, I didn't encounter any immediate warning or error messages. However, upon closer inspection of the Seurat object created, I noticed an unexpected issue.

After processing my data with the CreateSeuratObject function, the dimension names (Dimnames) of the dataset appear to be missing. This anomaly wasn't anticipated and isn't accompanied by any direct error messages, making it difficult to diagnose the problem. 
Additionally, I observed that the assay class in the Seurat object has changed to "Assay5". This is different from what I've seen typically and might be related to the issue at hand.

I've searched through the documentation and forums but haven't found a similar issue reported or addressed. I'm reaching out to see if anyone has encountered this problem or could provide insights into what might be causing the disappearance of dimnames and the change in the assay class.

I am attaching the screenshots for a clearer understanding. (Note: As per the guidelines, I'm not attaching the screenshots here but can provide them if needed.)

Here's a brief snippet of the code I used:

library(Seurat)
library(tidyverse)
library(dplyr)
library(patchwork)

scRNA.counts=Read10X("C:\\Users\\Admini\\Desktop\\51 lesson\\single cell\\lesson1\\GSE152048_BC21.matrix\\BC21")
scRNA = CreateSeuratObject(scRNA.counts,assay = "RNA",meta.data = NULL,project = "os", min.cells=3,min.features=300)
![屏幕截图 2023-12-28 155332](https://github.com/satijalab/seurat/assets/143380773/9435433f-0256-45f8-864d-f5ec07bed1c8)
![屏幕截图 2023-12-28 154429](https://github.com/satijalab/seurat/assets/143380773/16a18132-f9c9-45d8-9bb3-f2341807e9f0)

I would greatly appreciate any guidance or suggestions you might have. Thank you very much for your time and help!

Best Regards,
Tong Ling

**Comment by lian14697** [comment]
Is the R you using version 4.3?  please go back to version 4.2.2 and try again

**Reply by jarbaas** [reply]
Still not working on 4.2.2

**Comment by JordiBaRa** [comment]
I encountered the same issue. Not only I loose the dimnames, but counts are stored as a layer within RNA, but it used to be an independent feature within RNA. While it doesn't really affect the generation of UMAPs, it was problematic for me when performing DEseq analyses. 

I solved the problem by installing Seurat 4.4.0 and SeuratObject 4.1.4. 

However, is there any way this information could be retrieved from a Seurat Object in the newest version?

Thanks

**Reply by krasole** [reply]
I'm having the same issue, where no gene or cell names are stored in dimnames with seurat v5. I believe this is causing an issue with my ability to run AddModuleScore. I understand others have been able to resolve the latter issue with JoinLayers, but it has not worked for me.

**Reply by jarbaas** [reply]
Also having the same problem, can't do DESeq on my data and can't find a way to put the gene names back in 'counts'

**Reply by JordiBaRa** [reply]
Since Dimnames are still present in the "counts" object and are lost upon creating a Seurat Object, I solved it this way:

(Your seurat object)@assays[["RNA"]]@layers[["counts"]]@Dimnames<-(Your counts file)@Dimnames

You'd basically be reintroducing the same dimnames you have in the counts matrix back to the lost Dimnames layer in your Seurat Object. However, the resulting Seurat Object must have the same dimensions as your counts file. That means you cannot sort out features below your threshold settings on min.cells and min.features when creating the Seurat Object.

**Reply by jarbaas** [reply]
> Since Dimnames are still present in the "counts" object and are lost upon creating a Seurat Object, I solved it this way:
> 
> (Your seurat object)@assays[["RNA"]]@layers[["counts"]]@Dimnames<-(Your counts file)@Dimnames
> 
> You'd basically be reintroducing the same dimnames you have in the counts matrix back to the lost Dimnames layer in your Seurat Object. However, the resulting Seurat Object must have the same dimensions as your counts file. That means you cannot sort out features below your threshold settings on min.cells and min.features when creating the Seurat Object.

If my files are from a public database in h5 format is there still a way to do this? I don't have a counts matrix.

**Reply by AAVVKK** [reply]
This solution works, thank you.  But now - is there a way to drop empty cells and non-expressed genes from the Seurat object? Get the same result as would be obtained using min.features and min.cells parameters in CreateSeuratObject, but after the object has already been created?  Thanks!

**Reply by Jiankun-Liang** [reply]
Yes, I've encountered the same issue, But you can solve this by adding dimnames
#Create the data object
Obj <- CreateSeuratObject(counts = counts1, meta.data = meta, project = "Obj _bulk")

#solve seurat v5.0.1 Dimnames missing issue
MS1@assays$RNA@layers$counts@Dimnames <- list(rownames(counts1), colnames(counts1))

**Reply by mknewline** [reply]
Hi, I am in the same situation, feels like Seurat v5 is a step back...
However, I solved this issue by trimming data from the initial count matrix, then creating the Seurat Object.
What I did is the following:
**for "min.cell=3"** 
_index<- as.vector(rowSums(count_matrix[ , ]>0)>=3)_

**for "min.features=100"**
_index2 <- as.vector(colSums(count_matrix[ , ]>0)>=100)_

then trim matrix
_count_matrix <- count_matrix[index1,]_
_count_matrix <- count_matrix[,index2]_

Finally proceed to create object by setting min.cell and min.features to 0 and add back dimnames

seu_obj<- Seurat::CreateSeuratObject(counts = count_matrix,
                                       project = "project_name",
                                       min.cells = 0,
                                       min.features = 0, )

seu_obj@assays$RNA@layers$counts@Dimnames <- count_matrix@Dimnames

**Reply by JordiBaRa** [reply]
> Hi, I am in the same situation, feels like Seurat v5 is a step back... However, I solved this issue by trimming data from the initial count matrix, then creating the Seurat Object. What I did is the following: **for "min.cell=3"** _index<- as.vector(rowSums(count_matrix[ , ]>0)>=3)_
> 
> **for "min.features=100"** _index2 <- as.vector(colSums(count_matrix[ , ]>0)>=100)_
> 
> then trim matrix _count_matrix <- count_matrix[index1,]_ _count_matrix <- count_matrix[,index2]_
> 
> Finally proceed to create object by setting min.cell and min.features to 0 and add back dimnames
> 
> seu_obj<- Seurat::CreateSeuratObject(counts = count_matrix, project = "project_name", min.cells = 0, min.features = 0, )
> 
> seu_obj@assays$RNA@layers$counts@Dimnames <- count_matrix@Dimnames

This looks like an interesting solution, thank you!

**Comment by Clara22** [comment]
I found the 'old' Dimnames are now stored in obj@assays$RNA@cells and obj@assays$RNA@features, I guess it is trying to save space

**Reply by marzeeea** [reply]
They are accessible only via rownames() though, that is how I managed to add each of them into Dimnames

**Reply by lidiaga** [reply]
@Clara22 do your obj@assays$RNA@cells and obj@assays$RNA@features have names on them? Mine are logical vector filled with TRUE:

![imagen](https://github.com/user-attachments/assets/5e7eb8c6-1b9a-404e-9755-511be5825391)

**Comment by Malus-pumila** [comment]
I solved it this way:
seuratobj@assays$RNA@layers$data@Dimnames <- dimnames(seuratobj)
seuratobj@assays$RNA@layers$counts@Dimnames <- dimnames(seuratobj)
